plugins/net_plugin/net_plugin.cpp

```diff
@@ -3685,6 +3790,68 @@ namespace eosio {
       cc.process_vote_message(connection_id, msg);
    }
 
+   // called from connection strand
+   void connection::handle_message( const block_nack_message& msg ) {
+      auto block_num = block_header::num_from_id(msg.id);
+
+      peer_elog(this, "received block nack #${bn}:${id}, consecutive ${c}", ("bn", block_num)("id", msg.id)("c", consecutive_blocks_nacks));
+
+      if (block_num == 0) { // peer requested reset
+         consecutive_blocks_nacks = 0;
+         last_block_nack = msg.id;
+         return;
+      }
+
+      latest_blk_time = std::chrono::steady_clock::now();
+      auto fork_db_root_num = my_impl->get_fork_db_root_num();
+      const bool before_lib = block_header::num_from_id(msg.id) <= fork_db_root_num;
+
+      if (before_lib || my_impl->dispatcher.have_block(msg.id)) {
+         if (block_num - 1 == block_header::num_from_id(last_block_nack)) {
+            ++consecutive_blocks_nacks;
+         }
+         if (!before_lib) {
+            my_impl->dispatcher.add_peer_block(msg.id, connection_id);
+         }
+      }
+      last_block_nack = msg.id;
+   }
+
+   // called from connection strand
+   void connection::handle_message( const block_notice_message& msg ) {
+      auto fork_db_root_num = my_impl->get_fork_db_root_num();
+      if (block_header::num_from_id(msg.id) <= fork_db_root_num)
+         return;
+
+      latest_blk_time = std::chrono::steady_clock::now();
+      if (my_impl->dispatcher.have_block(msg.id)) {
+         my_impl->dispatcher.add_peer_block(msg.id, connection_id);
+      } else {
+         if (!last_block_notice.empty() && !my_impl->dispatcher.have_block(last_block_notice)) { // still don't have previous block
+            if (block_header::num_from_id(last_block_notice) == block_header::num_from_id(msg.id) - 1) {
+               peer_dlog(this, "Received 2 unknown block notices, see if already requested");
+               request_message req;
+               req.req_blocks.mode = normal;
+               req.req_blocks.ids.push_back(last_block_notice);
+               bool already_requested = my_impl->connections.any_of_block_connections([&req](const auto& c) {
```

Comment:
Since we haven't received it, even if we already requested the block from a different connection (whose host maybe is not responsive), probably we should request it again from the connection that sent us the `block notices`, since it is proven to be live and have the block.

Reply:
I added this because in practice is resulted in a storm of `request_message`s to many peers. The first received should be the best peer as you received the notice from it first.